### ISSUE TITLE
Filtering knapsackpro records

### ISSUE THREAD

**Issue by johngluckmdsol** [KnapsackPro/knapsack_pro-ruby#7]
We were running our tests with a cucumber tag, which was preventing some of our tests from getting noticed and timed by knapsack. This was a good thing.

Accidentally, we ended up running all our tests in travis and knapsackpro timed them all. Subsequent runs with the tag are resulting in knapsackpro returning a list of tests it wants to run that were not tagged, which means, no tests run in some nodes and travis fails for the node.

In order to reset, do we just need to get a new token? Is there way to tell knapsackpro to stop using a particular set of tests?

I should add here that I thought about using KNAPSACK_PRO_TEST_FILE_PATTERN because all the tests we want to run are in a subdir, but we are running both rspec and cucumber tests. I took a look at the code and, as far as I can tell, if I set this for cucumber, rspec will pick it up as well and fail because it won't find rspec files in the directory.

**Comment by johngluckmdsol** [comment]
BTW, the error we get when knapsackpro attempts to distribute tests to a node that aren't in the tag is

[knapsack_pro] API response:
[knapsack_pro] {"errors"=>[{"test_files"=>["parameter is required"]}]}
/home/travis/build/mdsol/parley/vendor/bundle/ruby/2.2.0/gems/knapsack_pro-0.7.0/lib/knapsack_pro/report.rb:15:in `save': {"errors"=>[{"test_files"=>["parameter is required"]}]} (ArgumentError)

**Comment by ArturT** [comment]
Could you add the example of commands you are using to run tests for rspec and cucumber?

You should use different API token for rspec and for cucumber test suite.

Don't set KNAPSACK_PRO_TEST_FILE_PATTERN as global var. Instead, add KNAPSACK_PRO_TEST_FILE_PATTERN as a prefix to command responsible for running rspec and cucumber tests.

```
# Step for RSpec
$ KNAPSACK_PRO_TEST_FILE_PATTERN="directory_with_specs/**{,/*/**}/*_spec.rb" KNAPSACK_PRO_CI_NODE_TOTAL=2 KNAPSACK_PRO_CI_NODE_INDEX=0 bundle exec rake knapsack_pro:rspec

# Step for Cucumber
$ KNAPSACK_PRO_TEST_FILE_PATTERN="directory_with_features/**{,/*/**}/*.feature" KNAPSACK_PRO_CI_NODE_TOTAL=2 KNAPSACK_PRO_CI_NODE_INDEX=0 bundle exec rake knapsack_pro:cucumber

```

**Comment by johngluckmdsol** [comment]
```
- bundle exec rake knapsack_pro:rspec -r ./spec/spec_helper.rb
- bundle exec rake "knapsack_pro:cucumber[--tag @Validation]"
```

**Comment by johngluckmdsol** [comment]
I'll try that.  Thanks

**Comment by ArturT** [comment]
bundle exec rake "knapsack_pro:cucumber[--tag @Validation]"

The command as a first thing collects all existing test files matching test file pattern and send them to API in order to detect later if you created a new tests files that should be split across nodes. I guess you have cucumber test files in different directory than default `features/**{,/*/**}/*.feature`. This should help:

KNAPSACK_PRO_TEST_FILE_PATTERN=="directory_with_features/**{,/*/**}/*.feature" bundle exec rake "knapsack_pro:cucumber[--tag @Validation]"

Please let me know if you solved the problem.

**Comment by johngluckmdsol** [comment]
So this worked better,but it still isn't giving us what we expect.  If I try this with two nodes and one test that is tagged @Test and I pass [--tag @Test], I will get one node where no tests are run and one node where all the tests are run.  Let me know if this is not clear

**Comment by ArturT** [comment]
> In other word, tags only work if the tag is found in each set of files that knapsack returns to each node.

This is expected behavior. You have only one test file with tag @Test. Knapsack Pro will split only test files across node (not particular test cases in test files). SIngle test file cannot be splited across 2 nodes. If your total number of nodes is 2 then you need to have at least 2 test files in order to split them.

For now you can try to divide your single feature test file in two smaller test files and then they can be distributed across nodes.

**Comment by johngluckmdsol** [comment]
I understand that it's working as designed.  I'm suggesting it's unintuitive.  I don't want my tests to fail simply because there weren't enough tests to distribute across the nodes.  For example, if I have six nodes and I select a tag that happens to have 5 tests that meet the criteria, I don't want my entire test run to fail because I didn't have enough tests to meet the criteria for six nodes.  That's what's happening now.

Furthermore, I don't want knapsack to revert to cucumber's default if it can't find anything.  I just want it to give up happily.  Right now, if it can't find anything on a particular node, it appears that it makes an attempt to default.  This is also incorrect in my mind.

**Comment by ArturT** [comment]
Hmm. What do you think we could do? Maybe on the node where there are no tests we could just show info to STDOUT that there are not tests to run and that's it? I can work on this fix tomorrow.

**Comment by johngluckmdsol** [comment]
Honestly, that is preferable to me.  I'm not sure about how everyone else using it feels but my guess is that this is an edge case.  That said, it doesn't feel like a failure to me.

**Comment by johngluckmdsol** [comment]
I thought about this and I'm wondering if there is some way to make the requested behavior optional

**Comment by ArturT** [comment]
I've just realised the error you saw was at the end of running tests. When no tests were run on node then knapsack_pro tried to submit empty list of test files for particular node to the API. I did fix to skip submitting empty list of test files but it's on branch `handle-less-tests` for now. Could you test it and let me know if it works for you?

    gem 'knapsack_pro', github: 'KnapsackPro/knapsack_pro-ruby', branch: 'handle-less-tests'

**Comment by ArturT** [comment]
@johngluckmdsol Did you try it? Is that working for you?

**Comment by johngluckmdsol** [comment]
Unfortunately, we're having trouble with our systems at the moment.  I'll let you know once I've been able to do it.

**Comment by johngluckmdsol** [comment]
@ArturT I've confirmed that this is much improved in that Knapsackpro is no longer failing.  However, I have 6 nodes, and one, and only, of the nodes ends up ignoring the specified tag.  I'm inclined to think this has something to do with knapsack_pro but I can't pinpoint why this is happening.  That said, this is improved.

```

**Comment by ArturT** [comment]
@johngluckmdsol Could you paste here the list of files that supposed to run on that particular node? Are those tests executed on the node (they were passed to cucumber command)? How are defined tests for that test with @Tag? Can you run tests locally?

**Comment by ArturT** [comment]
@johngluckmdsol I've released knapsack_pro 0.7.1 and removed the branch `handle-less-tests` so please use the knapsack_pro gem from rubygems.

I'm closing the story for now. If you find out something about the problem with ignoring specs you mentioned please create a new issue with details how to reproduce it or you can just drop me email (email is here http://knapsackpro.com) in case you don't want to share private data in the issue. I'm happy to help. :) Thanks for help to make knapsack_pro better!
